countItems.js

```diff
@@ -1,72 +1,165 @@
-const { MongoClientInterface } =
-    require('arsenal').storage.metadata.mongoclient;
+const async = require('async');
+const workerFarm = require('worker-farm');
 const { Logger } = require('werelogs');
-const fs = require('fs');
+const { reshapeExceptionError } = require('arsenal').errorUtils;
 
-const log = new Logger('S3Utils::ScanItemCount');
-const replicaSetHosts = process.env.MONGODB_REPLICASET;
-const database = process.env.MONGODB_DATABASE || 'metadata';
+const mongoClient = require('./countItems/mongoClient');
 
-function getIsLocationTransientCb() {
-    const locationConfigFile = 'conf/locationConfig.json';
-    if (!fs.existsSync(locationConfigFile)) {
-        log.info('location conf file missing, falling back to PENSIEVE coll',
-            { filename: locationConfigFile });
-        return null;
-    }
+const log = new Logger('S3Utils::CountItems::Master');
 
-    const buf = fs.readFileSync(locationConfigFile);
-    const locationConfig = JSON.parse(buf.toString());
+const numWorkers =
+    process.env.NUM_WORKERS && !Number.isNaN(process.env.NUM_WORKERS)
+    ? Number.parseInt(process.env.NUM_WORKERS, 10)
+    : 1;
 
-    return function locationIsTransient(locationName, log, cb) {
-        if (!locationConfig[locationName]) {
-            log.error('unknown location', { locationName });
-            process.nextTick(cb, null, false);
-            return;
-        }
-        const isTransient = Boolean(locationConfig[locationName].isTransient);
-        process.nextTick(cb, null, isTransient);
-    };
+let concurrentCursors = 10;
+if (process.env.CONCURRENT_CURSORS &&
+    !Number.isNaN(process.env.CONCURRENT_CURSORS)) {
+    concurrentCursors = Number.parseInt(process.env.CONCURRENT_CURSORS, 10);
 }
 
-const params = {
-    replicaSetHosts,
-    database,
-    isLocationTransient: getIsLocationTransientCb(),
-    writeConcern: 'majority',
-    replicaSet: 'rs0',
-    readPreference: 'secondaryPreferred',
-    replicationGroupId: 'RG001',
-    logger: log,
-};
+const workers = workerFarm(
+    {
```

Comment:
Don't these parameters cause 100 concurrent iterations? How was resource usage during tests?

Reply:
`worker-farm` does have a queue, and the number of concurrent open cursors will be `numWorkers * concurrent cursors`.
resource usage-wise, with 100 concurrent open cursors, mongo is running at 200-300%

Reply:
That seems a bit high, maybe confirm with @rahulreddy what the acceptable resource usage would be.

Reply:
resource usage wise:
there is a positive correlation between:
* number of cursors and the cpu usage of each worker
* number of workers and the cpu usage of mongo

remeasuring the time it takes to finish and the resources used